### ISSUE TITLE
[FR] Improve performance of plotting heatmaps

### ISSUE THREAD

**Issue by BioTurboNick** [JuliaPlots/Plots.jl#4520]
Reported on Slack by Jesse Chan, repeated calls to `plot!` to produce overlying heatmaps was very slow.

```
const mat = randn(512, 512)
function f(x)
    plot(); for _ in 1:10000
        heatmap!(x)
    end
end
```

Allocation profiler shows the large majority of allocations occurring in `heatmap_edges`:
![image](https://user-images.githubusercontent.com/1438610/202042099-e8caf599-bec0-49f3-826f-5f6fb16daecb.png)

And the large majority of computation time in `expand_extrema!` (which calls `heapmap_edges`) and `update_clims` (my old nemesis):
![image](https://user-images.githubusercontent.com/1438610/202042301-80dcc98e-05d0-4114-8b10-ee8c5a845617.png)

**Comment by t-bltg** [comment]
@jlchan, can you check if `Plots#master` improves the current or if it is still unacceptably slow ?

**Comment by jlchan** [comment]
Thanks @t-bltg. Using @BioTurboNick's snippet above, on Plots v1.36.1, I get
```julia
julia> @btime f($mat)
  511.680 ms (290431 allocations: 215.73 MiB)
```
On `Plots#master`
```julia
julia> @btime f($mat)
  509.745 ms (81638 allocations: 207.71 MiB)
```

**Comment by t-bltg** [comment]
Thanks, it's disappointing.
I'll give it another try later, memory allocation is just insane on my end.

**Comment by BioTurboNick** [comment]
Thanks for tackling this! I was going to poke around this week/weekend. Can still if you get stumped.

**Comment by jlchan** [comment]
I appreciate it! FYI I've found a workaround by adding another custom recipe. However, improving multiple `plot!` times would certainly help when I teach Julia to students who are coming from Matlab.

**Comment by t-bltg** [comment]
> Can still if you get stumped.

There is still room !
I'm filling up 64Gb RAM here, and starting to swap with that example.
Don't know why gc doesn't jump in though :thinking:.

**Comment by BioTurboNick** [comment]
Haha sorry. My system has 48 GB of RAM and caps out at 21 GB consumed by Julia. But you could easily cut the number down, Or cut the matrix size by half, too. I just chose a reasonably large one to show the effect, and I didn't pay attention to RAM usage haha.

Manually running GC doesn't reduce the number here.

Just holding on to all the plot data looks like 19.5 GB of RAM.

**Comment by t-bltg** [comment]
Let's focus on `n = 100` as of now (I've edited the examples).

The problem in this loop is that we accumulate series, so when exporting to `png` or `display`ing, we traverse `100` series.
This doesn't make sense when using `heatmap`s.

I've added the ability to empty the `Plot`'s series in https://github.com/JuliaPlots/Plots.jl/pull/4543.

So now the example looks like:
```julia
using BenchmarkTools, Plots

const mat = randn(512, 512)

f(x, n) = begin
  pl = plot()
  for _ in 1:n
      empty!(pl)  # clear out previous series
      heatmap!(x)
  end
  pl
end

@btime f($mat, 100)  #  597.178 ms (42561 allocations: 206.52 MiB)
```

It still allocates too much, so we can now focus on `Plots` internals to speed things up.

**Comment by BioTurboNick** [comment]
The motivation is actually to allow accumulating series here. They don't overlap in his use case, the example is just to emulate generating many series.

**Comment by BioTurboNick** [comment]
My suspicion is the problem is generic to number of series and size of data per series.

**Comment by jlchan** [comment]
Yeah, as @BioTurboNick said, I observe a similar phenomena when plotting many series over each other. 

An example of where I usually see this is creating a scatter plot from a matrix 
```julia
x = randn(100,1000)
y = randn(100,1000)
scatter(x, y)
```
This takes a lot longer than creating a scatter plot from `vec(x), vec(y), zcolor=vec(z)`, and I think this is because a different plot series is created for each column. 

My other use cases are similar; I'll tend to loop over each column of `x`

**Comment by jlchan** [comment]
I can add a MWE involving triangulations if that's helpful, but as @BioTurboNick said I don't think the issue is unique to `heatmap` plots.
